File: sample_repo/services/validation.py
"""Input validation for the signup payload."""

import re

EMAIL_PATTERN = re.compile(r"^[^@\s]+@[^@\s]+\.[a-zA-Z]{2,}$")
MIN_PASSWORD_LENGTH = 10


class ValidationError(Exception):
    """Raised when a signup payload fails validation."""


def validate_signup_input(payload: dict) -> dict:
    """Validate and normalise a signup payload."""
    email = check_email(payload.get("email", ""))
    password = check_password(payload.get("password", ""))
    display_name = normalise_name(payload.get("display_name", ""))
    return {"email": email, "password": password, "display_name": display_name}


def check_email(email: str) -> str:
    if not EMAIL_PATTERN.match(email.strip().lower()):
        raise ValidationError("invalid email address")
    return email.strip().lower()


def check_password(password: str) -> str:
    if len(password) < MIN_PASSWORD_LENGTH:
        raise ValidationError("password too short")
    return password


def normalise_name(display_name: str) -> str:
    cleaned = display_name.strip()
    return cleaned if cleaned else "anonymous"

File: sample_repo_ts/src/services/validation.ts
const EMAIL = /^[^@\s]+@[^@\s]+\.[a-zA-Z]{2,}$/;
const MIN_PASSWORD_LENGTH = 10;

export class ValidationError extends Error {}

export function checkEmail(email: string): string {
  const cleaned = email.trim().toLowerCase();
  if (!EMAIL.test(cleaned)) {
    throw new ValidationError("invalid email address");
  }
  return cleaned;
}

export function checkPassword(password: string): string {
  if (password.length < MIN_PASSWORD_LENGTH) {
    throw new ValidationError("password too short");
  }
  return password;
}

export function normaliseName(displayName: string): string {
  const cleaned = displayName.trim();
  return cleaned ? cleaned : "anonymous";
}

export function validateSignupInput(payload: Record<string, string>) {
  const email = checkEmail(payload.email ?? "");
  const password = checkPassword(payload.password ?? "");
  const displayName = normaliseName(payload.displayName ?? "");
  return { email, password, displayName };
}
